compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -484,46 +484,86 @@ class Namer { typer: Typer =>
-      for (cdef @ TypeDef(name, _) <- stats)
-        if (cdef.isClassDef) {
-          classDef(name) = cdef
-          cdef.attachmentOrElse(ExpandedTree, cdef) match {
-            case Thicket(cls :: mval :: (mcls @ TypeDef(mname, _: Template)) :: crest)
-            if name.moduleClassName == mname =>
-              moduleDef(name) = mcls
-            case _ =>
-          }
-        }
-      for (mdef @ ModuleDef(name, _) <- stats if !mdef.mods.is(Flags.Package)) {
-        val typName = name.toTypeName
-        // Expansion of object is a flattened thicket with the first two elements being:
-        //    module val :: module class :: rest
-        val Thicket(vdef :: (mcls @ TypeDef(_, impl: Template)) :: rest) = expanded(mdef)
-        moduleDef(typName) = mcls
-        classDef get name.toTypeName match {
-          case Some(cdef) =>
-            cdef.attachmentOrElse(ExpandedTree, cdef) match {
-              case Thicket(cls :: mval :: TypeDef(mname, compimpl: Template) :: crest)
-              if name.moduleClassName == mname =>
-                val mcls1 = cpy.TypeDef(mcls)(
-                    rhs = cpy.Template(impl)(body = compimpl.body ++ impl.body))
-                mdef.putAttachment(ExpandedTree, Thicket(vdef :: mcls1 :: rest))
-                moduleDef(typName) = mcls1
-                cdef.putAttachment(ExpandedTree, Thicket(cls :: crest))
+      def valid(mdef: MemberDef): Boolean = !mdef.mods.is(Package) && mdef.mods.is(Module)
+
+      for (stat <- stats)
+        expanded(stat) match {
+          case Thicket(trees) => // companion object always expands to thickets
+            trees.map {
+              case mcls @ TypeDef(name, impl: Template) if valid(mcls) =>
+                if (moduleClsDef.contains(name)) {
+                  val (stat1, mcls1 @ TypeDef(_, impl1: Template)) = moduleClsDef(name)
+                  if (mcls.mods.is(Synthetic) && !mcls1.mods.is(Synthetic)) { // merge synthetic into user-defined module
+                    removeInExpanded(stat, mcls)
+                    val mcls2 = mergeModuleClass(stat1, mcls1, impl.body)
+                    moduleClsDef(name) = (stat1, mcls2)
+                  }
+                  else if (!mcls.mods.is(Synthetic) && mcls1.mods.is(Synthetic)) {
+                    removeInExpanded(stat1, mcls1)
+                    val mcls2 = mergeModuleClass(stat, mcls, impl1.body)
+                    moduleClsDef(name) = (stat, mcls2)
+                  }
+                  else {
+                    // redefinition of objects or case classes, handled elsewhere
+                  }
+                }
+                else moduleClsDef(name) = (stat, mcls)
+
+              case vdef @ ValDef(name, _, _) if valid(vdef) =>
+                if (moduleValDef.contains(name)) {
+                  val (stat1, vdef1) = moduleValDef(name)
+                  if (vdef.mods.is(Synthetic) && !vdef1.mods.is(Synthetic))  // merge synthetic into user-defined module
```

Comment:
The comments are addressed in the latest commit.

For the last one, I think it's probably good to keep it as it is, as the processing is not symmetric as the case above, and the code is simple and short. Refactoring into a method seems to make readability worse.